src/proxy/inbound.rs

```diff
@@ -98,13 +103,21 @@ impl Inbound {
                 .http2_initial_stream_window_size(self.cfg.window_size)
                 .http2_initial_connection_window_size(self.cfg.connection_window_size)
                 .http2_max_frame_size(self.cfg.frame_size)
-                .serve(service);
+                .serve(service)
+                .with_graceful_shutdown(async {
+                    let shutdown = self.drain.signaled().await;
+                    // TODO: add some more graceful shutdown way
```

Comment:
can we comment why we do this? I agree it makes sense but its not intuitive. I think just saying something like

"We need 2 tiers of draining. First, we still accept new connections and just wait. This is to let Kubernetes propagate that we are shutting down and start directing new connections to another instance. Then we stop accepting connections and wait for existing ones to shut down"

This also probably belongs at a higher level; we don't want each component to try to hardcode the delay. Instead in `app.rs` Bound I would put the sleep between `shutdown.wait()` and `drain_tx.drain()` (and again, make it configurable)

Reply:
There should be some way to notify connection close, like close with a timeout.